README.md
# node-fast: streaming JSON RPC over TCP

Fast is a simple RPC protocol used in Joyent's
[Triton](http://github.com/joyent/triton) and
[Manta](https://github.com/joyent/manta) systems, particularly in the
[Moray](https://github.com/joyent/moray) key-value store.  This README contains
usage notes.  For developers, see CONTRIBUTING.md.

This module includes:

* client library interface
* server library interface
* `fastcall`, a command-line tool for making Fast RPC requests
* `fastserve`, a command-line Fast server for demo and testing
* `fastbench`, a command-line tool for very basic client load generation

This rewrite was done to fix a number of issues with service discovery and
connection management in the previous node-fast client.  Service discovery and
connection management in this client are managed by the caller using something
like [cueball](https://github.com/arekinath/node-cueball).

The library interfaces here follow Joyent's [Best Practices for Error
Handling](https://www.joyent.com/developers/node/design/errors).  In particular,
all operational errors are emitted asynchronously.  It is a programmer error to
call any of the public APIs with missing or incorrectly-typed arguments.  All
arguments are required unless otherwise specified.


## Synopsis

Start the Fast server:

    $ fastserve -p 2030 > server.log &

Use `fastcall` to invoke the `date` RPC method inside the client:

    $ fastcall 127.0.0.1 2030 date '[]'
    {"timestamp":1457475515355,"iso8601":"2016-03-08T22:18:35.355Z"}

Or try the `yes` method, an RPC version of yes(1):

    $ fastcall 127.0.0.1 2030 yes '[ { "value": { "hello": "world" }, "count": 3 } ]'
    {"hello":"world"}
    {"hello":"world"}
    {"hello":"world"}

For an example client and server, see the [fastcall](bin/fastcall) and
[fastserve](bin/fastserve) programs.


## Caveats

The Fast protocol is intended for internal systems.  It does not support
authentication.  Neither the client nor server implementations are hardened
against byzantine behavior (e.g., denial-of-service attacks), though they are
expected to correctly handle all network-related failure modes and byzantine
input.

Previous implementations of the Fast protocol supported cancellation, but it was
dangerous to use with servers that did not support it, and there was no way to
tell if the server did support it.  As a result, this implementation does not
support cancellation of in-flight requests.  (There's an `abandon()` function in
the client API, but it only causes the request stream to fail.  The underlying
RPC continues executing and incoming messages are ignored.)

Early versions of the original node-fast module used a
[buggy](https://github.com/alexgorbatchev/node-crc/issues/29) CRC
implementation.  Since changing the CRC algorithm would require a flag day among
deployed components, this module continues to use the buggy CRC implementation.


## Observability

### Kang server

The Fast client and server provide functions suitable for use with
[kang](https://github.com/davepacheco/kang), a small library for exposing
debugging information over an HTTP API.

The server-side kang functions report:

* server-wide statistics about connections created, requests started,
  requests completed, and requests failed;
* per-connection state information (including time accepted and errors seen) and
  statistics about requests started, completed, and failed; and
* per-request state information (including time started)

These enable basic monitoring of server activity and health.  The connection and
request state information allows you to see which clients are connected, how
long they've been connected, and how long individual requests have been running,
which is often helpful in identifying leaked or hung requests.

The client-side kang functions report per-client and per-request state and
statistics.

The client and server only provide functions; you have to hook this up to a kang
server.  The built-in demo server ("fastserve") and benchmarking tool
("fastbench") demonstrate how to do that.

The Kang and metric collection (see below) functionality are served through the
same [restify](https://github.com/restify/restify) server in "fastserve," but
that is not required. Optionally, Kang can be served using the `knStartServer`
function provided by the Kang library.

### Metric Collection

The Fast client and server optionally collect some basic request metrics using
[node-artedi](https://github.com/joyent/node-artedi).

The metrics collected are:

* total request count
* latency of requests, implemented as a histogram

You can pass an artedi-style metric collector into the Fast server or client
constructor to enable metric collection. An example of how to do this for the
server is included in the built-in demo server ("fastserve") and the
benchmarking tool ("fastbench") includes a client example with metrics enabled.

### DTrace probes

The Fast client and server provide DTrace probes and scripts in the "bin"
directory that use these probes to show client and server activity.  **These
scripts, the probes, and their arguments may change over time.**

**Fast client probes:**

Probe name  | Event                                              | Arg0                    | Arg1                     | Arg2                     | Arg3
----------- | -------------------------------------------------- | ----------------------- | ------------------------ | ------------------------ | ----
`rpc-start` | Client begins issuing an RPC call                  | (int) Client identifier | (int) Message identifier | (string) RPC method name | (json) object with "rpcargs" and optional "timeout".
`rpc-data`  | Client receives 'data' message for outstanding RPC | (int) Client identifier | (int) Message identifier | (json) Received data     | -
`rpc-done`  | Client finishes processing RPC                     | (int) Client identifier | (int) Message identifier | (json) May contain "error" describing any error that occurred. | -

**Fast client scripts in "bin" directory:**

* fastclatency: trace latency of all client RPC requests.  Prints power-of-two
  histogram of request latency when the script exits.
* fastcsnoop: dump out client RPC activity (all "start", "data", and "done"
  events).

Note that the client identifier is only unique within a process, and the
message identifier is only unique for a given client.  Both are only unique
over a given period of time (i.e., client ids and message ids may be reused).
See the sample scripts for how to use these correctly.

**Fast server probes:**

Probe name     | Event                                            | Arg0                    | Arg1                     | Arg2                     | Arg3
-------------- | ------------------------------------------------ | ----------------------- | ------------------------ | ------------------------ | ----
`conn-create`  | Client connection created                        | (int) server identifier | (int) client identifier  | (string) client label    | -
`conn-destroy` | Client connection destroyed                      | (int) server identifier | (int) client identifier  | -                        | -
`rpc-start`    | Server starts processing an RPC call             | (int) server identifier | (int) client identifier  | (int) request identifier | (string) RPC method name
`rpc-done`     | Server finishes processing an RPC call           | (int) server identifier | (int) client identifier  | (int) request identifier | -

**Fast server script in "bin" directory:**

* fastssnoop: dump out server RPC activity (client connection create/destroy and
  rpc start/done).

Similar to the client, the server identifier is only unique within the process,
and the request identifier is only unique within the client.  See the sample
scripts for how to use these correctly.

The client and server also use bunyan for logging.  On systems with DTrace
support, you can use runtime log snooping (`bunyan -p`) to observe what the
client and server would be logging at the finest-grained log levels, even if you
haven't enabled those.


## Client API

Consumers of the client API are responsible for maintaining persistent
connections to the server.  The
[cueball](http://github.com/arekinath/node-cueball) module is recommended for
this purpose.  Clients pass connected sockets to the FastClient constructor:


### FastClient class

A FastClient attaches to a Node `net.Socket` object.  The FastClient makes RPC
calls by sending Fast messages over the socket and reading responses in the
form of Fast messages from the socket.

Constructor arguments:

Name              | Type         | Meaning
----------------- | ------------ | -------
`collector`       | object       | [artedi](https://github.com/joyent/node-artedi)-style metric collector
`metricLabels`    | object       | [artedi](https://github.com/joyent/node-artedi)-style metric labels to be added to every metric collected. Note: `rpcMethod` is a reserved label and will be ignored if included.
`log`             | object       | [bunyan](https://github.com/trentm/node-bunyan)-style logger
`transport`       | `net.Socket` | underlying TCP connection to the server
`nRecentRequests` | positive int | number of recent requests to track for debugging purposes

While consumers are responsible for handling socket errors, the FastClient will
watch the `transport` for `error` and `end` events in order to proactively fail
any outstanding requests.

Public methods:

* `rpc(args)`: initiate an RPC request
* `rpcBufferAndCallback(args)`: initiate an RPC request and buffer incoming data
* `request.abandon()`: abandon an RPC request
* `detach()`: detach client from underlying socket

This class emits `error` when there's a problem with the underlying socket
(other than an `error` emitted by the socket itself) that prevents any requests
from completing.  This would usually be a protocol error of some sort.


#### rpc(args): initiate an RPC request to the remote server

Named arguments:

Name               | Type         | Meaning
---------------    | ------------ | -------
`rpcmethod`        | string       | name of the RPC method to invoke on the server
`rpcargs`          | array        | JSON-serializable array of RPC call arguments
`timeout`          | integer      | (optional) milliseconds after which to abandon the request if it has not already completed.  The default is that there is no timeout.
`log`              | object       | (optional) bunyan logger for this request.  If not specified, a child logger of the client-level logger will be used.
`ignoreNullValues` | boolean      | (optional) if true, null data values will be accepted from the server and dropped.  These are otherwise considered a protocol error.

The return value is an object-mode stream that consumers use to interact with
the request.  Objects sent by the server to the client are made available via
this stream.  The stream emits `end` when the server successfully completes the
request.  The stream emits `error` when the server reports an error, or if
there's a socket error or a protocol error.  Consumers need not proactively
abandon requests that fail due to a socket error.

Keep in mind that with any distributed system, failure of an RPC request due to
a socket error, protocol error, network failure, or timeout does not mean that
the RPC did not complete successfully or even that it is not still running.  The
server may have successfully completed the request, or failed it for a different
reason, or may still be running it when a network error occurs.  Consumers must
keep this in mind in designing RPC protocols and responses to failure.

As with other Node streams, the request stream will emit exactly one `end` or
`error` event, after which no other events will be emitted.


#### rpcBufferAndCallback(args, callback): initiate an RPC request and buffer response

This is a convenience function for making RPC calls when the server is expected
to return a bounded number of objects and the client doesn't intend to process
the results in a streaming way.  This function buffers up to
`maxObjectsToBuffer` objects and invokes `callback` when the request is
complete, as:

    callback(err, data, ndata)

where:

* `err` is the error emitted by the request
* `data` is an array of buffered data objects, which will have at most
  `maxObjectsToBuffer` elements
* `ndata` is a non-negative integer describing the total number of data objects
  received.  This may be larger than `data.length` only if some objects were
  dropped because `maxObjectsToBuffer` objects had already been buffered.

Note that `data` and `ndata` will always be present and valid, even if `err` is
non-null, since some number of data objects may have been received before the
request error.

Named arguments:

Name                 | Type                 | Meaning
-------------------- | -------------------- | -------
`rpcmethod`          | string               | see arguments to `rpc()`
`rpcargs`            | array                | see arguments to `rpc()`
`timeout`            | integer              | see arguments to `rpc()`
`log`                | object               | see arguments to `rpc()`

CHANGES.md
# Fast Changelog

## Not yet released.

None yet.

## v2.8.2

* [TOOLS-2525](https://jira.joyent.us/browse/TOOLS-2525) Everything needs to stop cloning with git:// URLs

## v2.8.1

* node-fast use wrong CRC library version when encoding errors to clients using
  updated CRC library version

## v2.8.0

 * node-fast provider transitive dependencies too old to build on x86_64

## v2.7.0

* #23 Add support for updated crc package to improve interoperability

## v2.6.1

* #22 Fix copy-paste bug in fast_server module

## v2.6.0

* `fast_client_request_time_ms` became `fast_client_request_time_seconds` and is
  now measured in seconds rather than milliseconds.
* `fast_request_time_ms` became `fast_server_request_time_seconds` and is
  now measured in seconds rather than milliseconds.
* Add support for artedi v2 which requires histograms to use fixed buckets.
  When the collector passed is an artedi v2 collector, default buckets are
  generated via artedi.logLinearBuckets(10, -1, 3, 5). See: joyent/node-artedi#17
  for more details. Data generated with artedi v1 collectors are likely
  invalid (and have been with previous releases as well). Dashboards/queries for
  services which pass v2 collectors to node-fast, should restrict results using
  the `{buckets_version="1"}` label when performing Prometheus queries on this
  data.

## v2.5.0

* #17 node-fast could track client metrics
* #16 fastserve cannot be run concurrently even with different arguments

## v2.4.0

* #14 want node-fast FastServer 'onConnsDestroyed' method

## v2.3.2

* #15 server leaks memory on socket errors
* #12 want support for node v6.12.0

## v2.3.1

* #11 add RPC method name to metrics

## v2.3.0

* #9 node-fast could track basic request metrics

## v2.2.4

* #7 fsr_context isn't always unpiped from fsr_encoder, which causes memory leaks

## v2.2.3

* #5 Server shutdown crashes when a connection had a socket error

## v2.2.2

* #4 client should preserve "ase\_errors" for Errors

## v2.2.1

* #3 client should not log at "error" level

## v2.2.0

* #2 update dependencies to support Node v4

## v2.1.0

* #1 need limited support for servers that send null values

## v2.0.0 (2016-08-29)

This is a complete rewrite of the Fast client and server implementation.  Almost
everything about the API has changed, including constructors, methods, and
arguments.  The primary difference is that the new Fast client is generally only
responsible for Fast-protocol-level concerns.  It does not do service discovery
(via DNS), connection establishment, health monitoring, retries, or anything
like that.  Callers are expected to use something like node-cueball for that.
The server manages a little bit more of the TCP state than the client does
(particularly as clients connect), but callers are still expected to manage the
server socket itself.  To summarize: callers are expected to manage their own
`net.Socket` (for clients) or `net.Server` (for servers), watching those objects
for whatever events they're interested in and using those classes' methods for
working with them.  You no longer treat the Fast client as a wrapper for a
`net.Socket` or the Fast server as a wrapper for a `net.Server`.

As a result of this change in design, the constructor arguments are pretty
different.  Many methods are gone (e.g., client.connect(), client.close(),
server.listen(), server.address(), and so on).  The interface is generally much
simpler.

To make methods more extensible and consistent with Joyent's Best Practices for
Error Handling (see README), the major APIs (constructors, `client.rpc()`, and
`server.rpc()`) have been changed to accept named arguments in a single `args`
argument.

Other concrete API differences include:

* Clients that make RPC calls get back a proper Readable stream, rather than an
  EventEmitter that emits `message`, `end`, and `error` events.  (The previous
  implementation used the spirit of Node.js's object-mode streams, but predated
  their maturity, and so implemented something similar but ad-hoc.) As a result
  of this change, you cannot send `null` over the RPC channel, since object-mode
  streams use `null` to denote end-of-stream.  Of course, you can wrap `null` in
  an object.
* `Client.rpc()` now takes named arguments as mentioned above.
* `Server.rpc()` is now called `Server.registerRpcMethod()` and takes named
  arguments as mentioned above.
* The server no longer provides an `after` event after RPC requests complete.

Other notable differences include:

* This implementation fixes a large number of protocol bugs, including cases
  where the protocol version was ignored in incoming messages and unsupported
  kinds of messages were treated as new RPC requests.
* This implementation does not support aborting RPC requests because the
  previous mechanism for this was somewhat dangerous.  (See README.)
* The client DTrace probes have been revised.  Server DTrace probes have been
  added.  (See README.)
* Kang entry points have been added for observability.  (See README.)

The original implementation can be found in the [v0.x
branch](https://github.com/joyent/node-fast/tree/fast-v0.x).


lib/fast_protocol.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright 2020 Joyent, Inc.
 */

/*
 * lib/fast_protocol.js: fast protocol definitions
 */

var mod_assertplus = require('assert-plus');
var mod_crc = require('crc');
var mod_extsprintf = require('extsprintf');
var mod_old_crc = require('oldcrc');
var mod_stream = require('stream');
var mod_util = require('util');
var VError = require('verror');

/* Exported interface */
exports.fastMessageEncode = fastMessageEncode;
exports.FastMessageEncoder = FastMessageEncoder;
exports.FastMessageDecoder = FastMessageDecoder;
/* Protocol constants are exported below. */

/*
 * Protocol definition
 *
 * All fast protocol messages look like the following:
 *
 *    0x00   +---------+--------+---------+---------+
 *           | VERSION | TYPE   | STATUS  | MSGID1  |
 *    0x04   +---------+--------+---------+---------+
 *           | MSGID2  | MSGID3 | MSGID4  | CRC1    |
 *    0x08   +---------+--------+---------+---------+
 *           | CRC2    | CRC3   | CRC4    | DLEN1   |
 *    0x0c   +---------+--------+---------+---------+
 *           | DLEN2   | DLEN3  | DLEN4   | DATA0   |
 *    0x10   +---------+--------+---------+---------+
 *           | DATAN...                             |
 *           +---------+--------+---------+---------+
 *
 * VERSION   1-byte integer.  The only supported values are "1" and "2".
 *
 * TYPE      1-byte integer.  The only supported value is TYPE_JSON (0x1),
 *           indicating that the data payload is an encoded JSON object.
 *
 * STATUS    1-byte integer.  The only supported values are:
 *
 *     STATUS_DATA  0x1  indicates a "data" message
 *
 *     STATUS_END   0x2  indicates an "end" message
 *
 *     STATUS_ERROR 0x3  indicates an "error" message
 *
 * MSGID1...MSGID4    4-byte big-endian unsigned integer, a unique identifier
 *                    for this message
 *
 * CRC1...CRC4        4-byte big-endian unsigned integer representing the CRC16
 *                    value of the data payload
 *
 * DLEN0...DLEN4      4-byte big-endian unsigned integer representing the number
 *                    of bytes of data payload that follow
 *
 * DATA0...DATAN      Data payload.  This is a JSON-encoded object (for TYPE =
 *                    TYPE_JSON).  The encoding length in bytes is given by the
 *                    DLEN0...DLEN4 bytes.
 *
 * Due to historical bugs in node-crc, the CRC implementation used in version 1
 * of the protocol is essentially incompatible with any CRC implementation other
 * than the one provided by node-crc version 0.x. The changes to address this
 * incompatibility (originally recorded on Github as joyent/node-fast#23) were
 * done to create a migration path for clients and servers off of the buggy
 * version with minimal operational impact. Clients and servers may select a
 * mode of operation where only the buggy CRC calculation is used or where only
 * a correct version of the calculation is used. Additionally, servers may elect
 * to operate in a mode where messages whose CRC is calculated with either the
 * buggy or correct implementation are accepted. In this mode the server will
 * use the same CRC calculation method in response messages to a client that the
 * client used when encoding the message to transmit to the server. This allows
 * for clients to be updated in a gradual manner to support the correct CRC
 * calculation. The FAST_CHECKSUM_* constants in this file control the mode of
 * operation with respect to the CRC calculation of clients and servers. This
 * mechanism can be reused in the event of other such library incompatibilities.
 * This mechanism is not intended or required for normal version upgrades to the
 * CRC library dependency where there is no change in the result produced by one
 * of the CRC calculation methods used by node-fast (crc16 is currently the only
 * one used).
 */

/*
 * Message IDs: each Fast message has a message id, which is scoped to the Fast
 * connection.  We allocate these sequentially from a circular 31-bit space.
 */
var FP_MSGID_MAX        = Math.pow(2, 31) - 1;
exports.FP_MSGID_MAX    = FP_MSGID_MAX;

/*
 * Field offsets
 */
var FP_OFF_VERSION      = 0x0;
var FP_OFF_TYPE         = FP_OFF_VERSION + 0x1;	/* 0x1 */
var FP_OFF_STATUS       = FP_OFF_TYPE + 0x1;	/* 0x2 */
var FP_OFF_MSGID        = FP_OFF_STATUS + 0x1;   /* 0x3 */
var FP_OFF_CRC          = FP_OFF_MSGID + 0x4;	/* 0x7 */
var FP_OFF_DATALEN      = FP_OFF_CRC + 0x4;	/* 0xb */
var FP_OFF_DATA         = FP_OFF_DATALEN + 0x4;	/* 0xf */
exports.FP_OFF_VERSION  = FP_OFF_VERSION;
exports.FP_OFF_TYPE     = FP_OFF_TYPE;
exports.FP_OFF_STATUS   = FP_OFF_STATUS;
exports.FP_OFF_MSGID    = FP_OFF_MSGID;
exports.FP_OFF_CRC      = FP_OFF_CRC;
exports.FP_OFF_DATALEN  = FP_OFF_DATALEN;
exports.FP_OFF_DATA     = FP_OFF_DATA;


/* size (in bytes) of each message header */
var FP_HEADER_SZ        = FP_OFF_DATA;		/* 0xf */
exports.FP_HEADER_SZ    = FP_HEADER_SZ;

/* possible values for the "status" byte */
var FP_STATUS_DATA      = 0x1;
var FP_STATUS_END       = 0x2;
var FP_STATUS_ERROR     = 0x3;
exports.FP_STATUS_DATA  = FP_STATUS_DATA;
exports.FP_STATUS_END   = FP_STATUS_END;
exports.FP_STATUS_ERROR = FP_STATUS_ERROR;

/* possible values for the "type" byte */
var FP_TYPE_JSON        = 0x1;
exports.FP_TYPE_JSON    = FP_TYPE_JSON;

/* possible values for the "version" byte */
var FP_VERSION_1           = 0x1;
var FP_VERSION_2           = 0x2;
var FP_VERSION_CURRENT     = FP_VERSION_2;
exports.FP_VERSION_1       = FP_VERSION_1;
exports.FP_VERSION_2       = FP_VERSION_2;
exports.FP_VERSION_CURRENT = FP_VERSION_CURRENT;

// These constants are facilitate an upgrade path from buggy node-crc@0.3.0
var FAST_CHECKSUM_V1         = 0x1;
var FAST_CHECKSUM_V1_V2      = 0x2;
var FAST_CHECKSUM_V2         = 0x3;
exports.FAST_CHECKSUM_V1     = FAST_CHECKSUM_V1;
exports.FAST_CHECKSUM_V1_V2  = FAST_CHECKSUM_V1_V2;
exports.FAST_CHECKSUM_V2     = FAST_CHECKSUM_V2;

/*
 * Encode a logical message for sending over the wire.  This requires the
 * following named properties:
 *
 *     msgid    (number) message identifier -- see lib/fast.js
 *
 *     data     (object) represents message contents.  At this level, this
 *                       can be any plain-old JavaScript object.
 *
 *     status   (number) message "status" (one of FP_STATUS_DATA, FP_STATUS_END,
 *                       or FP_STATUS_ERROR).
 *
 * Failure to match these requirements is a programmer error that may result in
 * a synchronously thrown exception that should not be caught.
 */
function fastMessageEncode(msg)
{
	var buffer, data_encoded, datalen, crc16;

	mod_assertplus.object(msg, 'msg');
	mod_assertplus.ok(typeof (msg.msgid) == 'number' &&
	    Math.floor(msg.msgid) == msg.msgid &&
	    msg.msgid >= 0 && msg.msgid <= FP_MSGID_MAX,
	    'msg.msgid is not an integer between 0 and FP_MSGID_MAX');
	mod_assertplus.object(msg.data, 'msg.data');
	mod_assertplus.number(msg.status, 'msg.status');
	mod_assertplus.number(msg.version, 'msg.version');

	switch (msg.status) {
	case FP_STATUS_DATA:
	case FP_STATUS_END:
	case FP_STATUS_ERROR:
		break;
	default:
		throw (new VError('unsupported fast message status'));
	}

	data_encoded = JSON.stringify(msg.data);

	/*
	 * Fast version 1 used a buggy version of the node-crc library so we
	 * special case it here. All subsequent versions default to using a
	 * corrected version of the node-crc library.
	 */
	if (msg.version === FP_VERSION_1) {
		crc16 = mod_old_crc.crc16(data_encoded);
	} else {
		crc16 = mod_crc.crc16(data_encoded);
	}

	datalen = Buffer.byteLength(data_encoded);
	buffer = new Buffer(FP_HEADER_SZ + datalen);
	buffer.writeUInt8(msg.version, FP_OFF_VERSION);
	buffer.writeUInt8(FP_TYPE_JSON, FP_OFF_TYPE);
	buffer.writeUInt8(msg.status, FP_OFF_STATUS);
	buffer.writeUInt32BE(msg.msgid, FP_OFF_MSGID);
	buffer.writeUInt32BE(crc16, FP_OFF_CRC);
	buffer.writeUInt32BE(datalen, FP_OFF_DATALEN);
	buffer.write(data_encoded, FP_OFF_DATA, datalen, 'utf8');
	return (buffer);
}

/*
 * Validate the message CRC based on the protocol version. Returns an
 * object that contains the decoded CRC mode if the validation is successful
 * or an error if the validation fails.
 *
 *     version		The protocol version in use. Used to transition off
 *             		buggy node-crc version. See the comments at the top
 *                      of this module for more details.
 *     headerCrc        The CRC provided in the Fast message header.
 *     data             The string representation of the Fast message data.
 */
function validateCrc(version, headerCrc, data) {
	var valid = true;
	var error;

	switch (version) {
	case FP_VERSION_1:
		var v1CalculatedCrc = mod_old_crc.crc16(data);
		if (v1CalculatedCrc !== headerCrc) {
			valid = false;
			error = crcValidationError(headerCrc, v1CalculatedCrc,
			    version);
		}
		break;
	default:
		var v2CalculatedCrc = mod_crc.crc16(data);
		if (v2CalculatedCrc !== headerCrc) {
			valid = false;
			error = crcValidationError(headerCrc, v2CalculatedCrc,
			    version);
		}
		break;
	}

	return ({ valid: valid, error: error});
}

/*
 * Return a Verror with details about a CRC validation failure.
 *
 *     headerCrc        The CRC provided in the Fast message header.
 *     calculatedCrc    The calculated CRC for the received message.
 *     version		The protocol version in use.
 */
function crcValidationError(headerCrc, calculatedCrc, version) {
	var errorObj = {
	    'name': 'FastProtocolError',
	    'info': {
		'fastReason': 'bad_crc',
		'crcExpected': headerCrc,
		'crcCalculated': calculatedCrc,
		'version': version
	    }
	};
	return (new VError(errorObj, 'fast protocol: expected CRC %s, found %s',
	    headerCrc, calculatedCrc));
}

/*
 * Decode a fast message from a buffer that's known to contain a single,
 * well-formed message.  All of the protocol fields are known to be valid (e.g.,
 * version, type, status, and msgid) at this point, but the data has not been
 * read, so the CRC has not been validated.
 */
function fastMessageDecode(header, buffer)
{
	var datalen, datastr, json;

	mod_assertplus.number(header.datalen, 'header.datalen');
	datalen = header.datalen;
	mod_assertplus.equal(buffer.length, FP_OFF_DATA + datalen);
	datastr = buffer.toString('utf8', FP_OFF_DATA);

	var crcValidationResult = validateCrc(header.version, header.crc,
	    datastr);

	if (crcValidationResult && !crcValidationResult.valid) {
		return (crcValidationResult.error);
	}

	try {
		json = JSON.parse(datastr);
	} catch (ex) {
		return (new VError({
		    'name': 'FastProtocolError',
		    'cause': ex,
		    'info': {
			'fastReason': 'invalid_json'
		    }
		}, 'fast protocol: invalid JSON in "data"'));
	}

	if (typeof (json) != 'object' || json === null) {
		return (new VError({
		    'name': 'FastProtocolError',
		    'info': {
			'fastReason': 'bad_data'
		    }
		}, 'fast protocol: message data must be a non-null object'));
	}

	if ((header.status == FP_STATUS_DATA ||
	    header.status == FP_STATUS_END) && !Array.isArray(json.d)) {
		return (new VError({
		    'name': 'FastProtocolError',
		    'info': {
			'fastReason': 'bad_data_d'
		    }
		}, 'fast protocol: data.d for DATA and END messages must be ' +
		    'an array'));
	}

	if (header.status == FP_STATUS_ERROR &&
	    (typeof (json.d) != 'object' || json.d === null ||
	    typeof (json.d.name) != 'string' ||
	    typeof (json.d.message) != 'string')) {
		return (new VError({
		    'name': 'FastProtocolError',
		    'info': {
			'fastReason': 'bad_error'
		    }
		}, 'fast protocol: data.d for ERROR messages must have name ' +
		    'and message'));
	}

	return ({
	    'status': header.status,
	    'msgid': header.msgid,
	    'data': json,
	    'version': header.version
	});
}

/*
 * Transform stream that takes logical messages and emits a buffer representing
 * that message (for sending over the wire).
 */
function FastMessageEncoder()
{
	mod_stream.Transform.call(this, {
	    'highWaterMark': 16,
	    'objectMode': true
	});
}

mod_util.inherits(FastMessageEncoder, mod_stream.Transform);

FastMessageEncoder.prototype._transform = function (chunk, _, callback)
{
	this.push(fastMessageEncode(chunk));
	setImmediate(callback);
};


/*
 * Transform stream that takes bytes (via Buffer objects) and emits an object
 * representing the encoded Fast message.
 */
function FastMessageDecoder()
{
	mod_stream.Transform.call(this, {
	    'objectMode': true
	});

	/* current state */
	this.md_buffer = null;		/* unparsed data */
	this.md_havebytes = 0;		/* bytes of unparsed data */
	this.md_done = false;		/* we've read end-of-stream */
	this.md_error = null;		/* fatal error */
	this.md_pushing = false;	/* currently calling push() */

	/* current header */
	this.md_version = null;
	this.md_type = null;
	this.md_status = null;
	this.md_msgid = null;
	this.md_crc = null;
	this.md_datalen = null;

	/* debug information */
	this.md_nmessages = 0;
	this.md_nbytes = 0;
}

mod_util.inherits(FastMessageDecoder, mod_stream.Transform);

FastMessageDecoder.prototype._transform = function (chunk, _, callback)
{
	this.md_havebytes += chunk.length;

	if (this.md_buffer === null) {
		mod_assertplus.equal(this.md_havebytes, chunk.length);
		this.md_buffer = chunk;
	} else {
		this.md_buffer = Buffer.concat(
		    [ this.md_buffer, chunk ], this.md_havebytes);
	}

	this.md_nbytes += chunk.length;
	this.decode(callback);
};

FastMessageDecoder.prototype._flush = function (callback)
{
	this.md_done = true;
	this.decode(callback);
};

FastMessageDecoder.prototype.decode = function (callback)
{
	var buf, msg;

	if (this.md_pushing) {
		return;
	}

	mod_assertplus.ok(this.md_error === null);

	while (this.md_havebytes >= FP_HEADER_SZ) {
		buf = this.md_buffer;
		mod_assertplus.equal(buf.length, this.md_havebytes);
		mod_assertplus.ok(buf !== null);
		mod_assertplus.ok(this.md_error === null);
		mod_assertplus.ok(this.md_version === null);

		this.md_version = buf.readUInt8(FP_OFF_VERSION);
		if (this.md_version != FP_VERSION_CURRENT &&
		    this.md_version != FP_VERSION_1) {
			this.md_error = new VError({
			    'name': 'FastProtocolError',
			    'info': {
				'fastReason': 'unsupported_version',
				'foundVersion': this.md_version
			    }
			}, 'fast protocol: unsupported version %d',
			    this.md_version);
			break;
		}

		this.md_type = buf.readUInt8(FP_OFF_TYPE);
		if (this.md_type != FP_TYPE_JSON) {
			this.md_error = new VError({
			    'name': 'FastProtocolError',
			    'info': {
			        'fastReason': 'unsupported_type',
				'foundType': this.md_type
			    }
			}, 'fast protocol: unsupported type 0x%x',
			    this.md_type);
			break;
		}

		this.md_status = buf.readUInt8(FP_OFF_STATUS);
		switch (this.md_status) {
		case FP_STATUS_DATA:
		case FP_STATUS_END:
		case FP_STATUS_ERROR:
			break;
		default:
			this.md_error = new VError({
			    'name': 'FastProtocolError',
			    'info': {
			        'fastReason': 'unsupported_status',
				'foundStatus': this.md_status
			    }
			}, 'fast protocol: unsupported status 0x%x',
			    this.md_status);
			break;
		}

		if (this.md_error !== null) {
			break;
		}

		this.md_msgid = buf.readUInt32BE(FP_OFF_MSGID);
		if (this.md_msgid < 0 || this.md_msgid > FP_MSGID_MAX) {
			this.md_error = new VError({
			    'name': 'FastProtocolError',
			    'info': {
				'fastReason': 'invalid_msgid',
				'foundMsgid': this.md_msgid
			    }
			}, 'fast protocol: invalid msgid %s', this.md_msgid);
			break;
		}

		this.md_crc = buf.readUInt32BE(FP_OFF_CRC);
		this.md_datalen = buf.readUInt32BE(FP_OFF_DATALEN);

		if (this.md_havebytes < FP_HEADER_SZ + this.md_datalen) {
			/*
			 * We don't have enough bytes to continue.  Stop now.
			 * We'll end up re-parsing the header again when we have
			 * more data.  If that turns out to be expensive, we can
			 * rework this code to keep track of where we were.
			 */
			this.md_version = null;
			this.md_type = null;
			this.md_status = null;
			this.md_msgid = null;
			this.md_crc = null;
			this.md_datalen = null;
			break;
		}

		/*
		 * We have a complete message.  Consume it and update our buffer
		 * state.
		 */
		buf = this.md_buffer.slice(0, FP_HEADER_SZ + this.md_datalen);
		this.md_buffer = this.md_buffer.slice(
		    FP_HEADER_SZ + this.md_datalen);
		this.md_havebytes -= buf.length;
		msg = fastMessageDecode({
		    'version': this.md_version,
		    'type': this.md_type,
		    'status': this.md_status,
		    'msgid': this.md_msgid,
		    'crc': this.md_crc,
		    'datalen': this.md_datalen
		}, buf);
		if (msg instanceof Error) {
			this.md_error = msg;
			break;
		}

		this.md_version = null;
		this.md_type = null;
		this.md_status = null;
		this.md_msgid = null;
		this.md_crc = null;
		this.md_datalen = null;

		this.md_pushing = true;
		this.push(msg);
		this.md_pushing = false;
		this.md_nmessages++;
	}

	if (this.md_error === null && this.md_havebytes > 0 && this.md_done) {
		this.md_error = new VError({
		    'name': 'FastProtocolError',
		    'info': {
			'fastReason': 'incomplete_message'
		    }
		}, 'fast protocol: incomplete message at end-of-stream');
	}

	if (this.md_error !== null) {
		setImmediate(callback, this.md_error);
	} else {
		setImmediate(callback);
	}
};
